_posts/2021-01-11-pisanie-pluginow-do-intellij.md

```diff
@@ -0,0 +1,324 @@
+
+#### Generowanie dokumentacji
+
+Dokumentacja będzie składać się z wierszy zwierających nazwę dokumentowanego elementu(np. nazwa pola) oraz opisu z adnotacji. Przyjmując, że na wejsciu mamy model PSI pliku z kodem, wygenerowanie takiej dokumentacji może wyglądać mniej więcej tak:
+
+```java
+
+    private String getDoc(PsiFile psiFile) {
+        if (psiFile instanceof PsiJavaFile) {
+            PsiJavaFile javaFile = (PsiJavaFile) psiFile;
+
+            Map<String, String> doc = Arrays.stream(javaFile.getClasses())
+                    .filter(psiClass -> psiClass.hasAnnotation("com.consdata.doc.Doc"))
+                    .collect(Collectors.toMap(PsiClass::getName, psiClass -> evaulate(psiClass.getProject(), psiClass.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value"))));
+
+            doc.putAll(Arrays.stream(javaFile.getClasses())
+                    .map(PsiClass::getFields)
+                    .flatMap(Arrays::stream)
+                    .filter(psiField -> psiField.hasAnnotation("com.consdata.doc.Doc"))
+                    .collect(Collectors.toMap(PsiField::getName, psiField -> evaulate(psiField.getProject(), psiField.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value")))));
+
+            doc.putAll(Arrays.stream(javaFile.getClasses())
+                    .map(PsiClass::getMethods)
+                    .flatMap(Arrays::stream)
+                    .filter(psiMethod -> psiMethod.hasAnnotation("com.consdata.doc.Doc"))
+                    .collect(Collectors.toMap(PsiMethod::getName, psiMethod -> evaulate(psiMethod.getProject(), psiMethod.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value")))));
+
+            return String.join("\n", doc.entrySet().stream().map(e -> e.getKey() + ": " + e.getValue()).collect(Collectors.toList()));
+        } else {
+            return "";
+        }
+    }
+
+```
+Korzystając z PSI przechodzimy po wszystkich klasach w pliku oraz ich polach i metodach. Jeśli zawierają adnotacją `Doc`, to do wierszy dokumentacji dodajemy nazwę pola, znak ':' oraz wartość adnotacji. Nie możemy jednak wartości adnotacji pobrać wprost. Nie jest to String. Zależnie od tego co znajduje się w adnotacji będzie to model jakiegoś wyrażenia. Musimy takie wyrażenie wyliczyć(aby np. rozwiązać użyte w nim stałe). Dla ułatwienia tego typu zadań powstał ewaluator stałych, który jest częścią fasady Javy dla PSI:
```

Comment:
"Jeśli zawierają adnotacją" -> "Jeśli zawierają adnotację" 
"Nie możemy jednak wartości adnotacji pobrać wprost. Nie jest to String." - sklejmy to w jedno zdanie, na przykład "Nie możemy [...], gdyż nie jest to String". 
Ponownie spacja przed otwarciem nawiasu